core/src/main/resources/META-INF/patches/groovy/02-fix-providerKey-constraints.resolved.groovy

```diff
@@ -0,0 +1,422 @@
+import com.fasterxml.jackson.databind.ObjectMapper
+import org.apache.commons.lang.StringUtils
+import org.apache.jackrabbit.core.JahiaRepositoryImpl
+import org.apache.jackrabbit.core.SearchManager
+import org.apache.jackrabbit.core.id.NodeId
+import org.jahia.osgi.BundleUtils
+import org.jahia.services.content.JCRCallback
+import org.jahia.services.content.JCRObservationManager
+import org.jahia.services.content.JCRSessionWrapper
+import org.jahia.services.content.JCRTemplate
+import org.jahia.services.content.impl.jackrabbit.SpringJackrabbitRepository
+
+import javax.jcr.Property
+import javax.jcr.PropertyIterator
+import javax.jcr.PropertyType
+import javax.jcr.RepositoryException
+import javax.sql.DataSource
+import java.sql.SQLException
+
+/**
+ * Find duplicate providers in the jahia_external_provider_id table
+ * @param dataSource
+ * @return a map of duplicated providerKey -> [id1, id2, ...]
+ */
+static def findDuplicateProviders(DataSource dataSource) {
+    def duplicates = [:].withDefault { [] }
+    def query = """
+        SELECT id, providerKey FROM jahia_external_provider_id 
+        WHERE providerKey IN (
+            SELECT providerKey FROM jahia_external_provider_id 
+            GROUP BY providerKey HAVING COUNT(*) > 1
+        )
+    """
+
+    try (def connection = dataSource.connection;
+         def statement = connection.prepareStatement(query);
+         def resultSet = statement.executeQuery()) {
+
+        while (resultSet.next()) {
+            def id = resultSet.getString("id")
+            def providerKey = resultSet.getString("providerKey")
+            duplicates[providerKey] << id
+        }
+    } catch (SQLException e) {
+        e.printStackTrace()
```

Comment:
Never just log to standard error, but use log.error. If there is an SQLException I don't think it is safe to assume that there are no duplicates.

Reply:
Yes better letting the error throw, if anything crash the Patcher will catch the error globally for the complete script execution.